From 4c515284a4bc9eccb0baf1c6863f8d76a12bf333 Mon Sep 17 00:00:00 2001
From: Vladimir Barinov <vladimir.barinov@cogentembedded.com>
Date: Wed, 9 Aug 2017 11:52:22 +0300
Subject: [PATCH 837/909] drm: adv7511: use smbus to retrieve edid

commit d8e286fd757432df78121151e5174ff64d902add from
https://github.com/CogentEmbedded/meta-rcar.git

Get EDID using smbus protocol instead block i2c transfer
This fixes often checksum errors while retriving EDID at 400kHz bus speed

Signed-off-by: Vladimir Barinov <vladimir.barinov@cogentembedded.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/bridge/adv7511/adv7511_drv.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/bridge/adv7511/adv7511_drv.c b/drivers/gpu/drm/bridge/adv7511/adv7511_drv.c
index 470be692ef09..06a70f02b912 100644
--- a/drivers/gpu/drm/bridge/adv7511/adv7511_drv.c
+++ b/drivers/gpu/drm/bridge/adv7511/adv7511_drv.c
@@ -553,18 +553,19 @@ static int adv7511_get_edid_block(void *data, u8 *buf, unsigned int block,
 		 * support 64 byte transfers than 256 byte transfers
 		 */
 
+#define CHUNK_SIZE	1
 		xfer[0].addr = adv7511->i2c_edid->addr;
 		xfer[0].flags = 0;
 		xfer[0].len = 1;
 		xfer[0].buf = &offset;
 		xfer[1].addr = adv7511->i2c_edid->addr;
 		xfer[1].flags = I2C_M_RD;
-		xfer[1].len = 64;
+		xfer[1].len = CHUNK_SIZE;
 		xfer[1].buf = adv7511->edid_buf;
 
 		offset = 0;
 
-		for (i = 0; i < 4; ++i) {
+		for (i = 0; i < 256/CHUNK_SIZE; ++i) {
 			ret = i2c_transfer(adv7511->i2c_edid->adapter, xfer,
 					   ARRAY_SIZE(xfer));
 			if (ret < 0)
@@ -572,8 +573,8 @@ static int adv7511_get_edid_block(void *data, u8 *buf, unsigned int block,
 			else if (ret != 2)
 				return -EIO;
 
-			xfer[1].buf += 64;
-			offset += 64;
+			xfer[1].buf += CHUNK_SIZE;
+			offset += CHUNK_SIZE;
 		}
 
 		adv7511->current_edid_segment = block / 2;
-- 
2.17.1

